contracts/near/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+            let vaa = vaa_payload::from_slice_with_payload::<structs::Vaa<()>>(&vaa)
+                .map_err(|_| Error::InvalidVaa)?;
+
+            if !self.source.contains(&Source {
+                emitter:            vaa.0.emitter_address.0,
+                pyth_emitter_chain: vaa.0.emitter_chain as u16,
+            }) {
+                return Err(Error::UnknownSource);
+            }
+        }
+
+        // Verify VAA and refund the caller in case of failure.
+        ext::ext_worm_hole::ext(env::current_account_id())
+            .with_static_gas(Gas(30_000_000_000_000))
+            .verify_vaa(vaa.clone())
+            .then(
+                Self::ext(env::current_account_id())
+                    .with_static_gas(Gas(30_000_000_000_000))
+                    .with_attached_deposit(env::attached_deposit())
+                    .process_vaa_callback(env::predecessor_account_id(), vaa),
+            )
+            .then(
+                Self::ext(env::current_account_id())
+                    .with_static_gas(Gas(30_000_000_000_000))
+                    .refund_vaa(env::predecessor_account_id(), env::attached_deposit()),
+            );
+
+        Ok(())
+    }
+
+    /// If submitting the VAA fails then this callback will refund the caller.
+    pub fn refund_vaa(&mut self, account_id: AccountId, amount: u128) {
+        if !is_promise_success() {
+            Promise::new(account_id).transfer(amount);
+        }
+    }
+
+    #[private]
+    pub fn set_upgrade_hash(&mut self, codehash: [u8; 32]) {
+        self.codehash = codehash;
+    }
+
+    /// Upon succesful verification of a VAA, this callback processes the VAA itself and branches
+    /// to the appropriate handler.
+    #[private]
+    pub fn process_vaa_callback(
+        &mut self,
+        account_id: AccountId,
+        vaa: String,
+        #[callback_result] _result: Result<u32, near_sdk::PromiseError>,
+    ) {
+        // Get Storage Usage before executed.
+        let storage_usage = env::storage_usage();
+
+        // Deserialize VAA, note that we already deserialized and verified the VAA in `process_vaa`
+        // at this point so we only care about the `rest` component which contains bytes we can
+        // deserialize into an Action.
+        let vaa = hex::decode(&vaa).unwrap();
+        let (_, rest): (structs::Vaa<()>, _) = vaa_payload::from_slice_with_payload(&vaa).unwrap();
+
+        match Action::try_from_slice(rest).unwrap() {
+            Action::AddSource(source) => self.add_source(source),
+            Action::DelSource(source) => self.del_source(source),
+            Action::UpgradeContract(codehash) => self.set_upgrade_hash(codehash),
+            Action::BatchAttest(batch) => {
+                let bytes = &mut Cursor::new(batch);
+                let batch = BatchPriceAttestation::deserialize(bytes).unwrap();
+                let count = self.attest_prices(&batch);
+                log!("Updated {} Price Feeds", count);
```

Comment:
It might be nice to emit event log messages: https://github.com/near/NEPs/blob/master/specs/Standards/EventsFormat.md